src/services/goToDefinition.ts

```diff
@@ -87,6 +87,39 @@ namespace ts.GoToDefinition {
                 declaration => createDefinitionInfo(declaration, shorthandSymbolKind, shorthandSymbolName, shorthandContainerName));
         }
 
+        if (isJsxOpeningLikeElement(node.parent)) {
+            // For JSX opening-like element, the tag can be resolved either as stateful component (e.g class) or stateless function component.
+            // Because if it is a stateless function component with an error while trying to resolve the signature, we don't want to return all
+            // possible overloads but just the first one.
+            // For example:
+            //      /*firstSource*/declare function MainButton(buttonProps: ButtonProps): JSX.Element;
+            //      /*secondSource*/declare function MainButton(linkProps: LinkProps): JSX.Element;
+            //      /*thirdSource*/declare function MainButton(props: ButtonProps | LinkProps): JSX.Element;
+            //      let opt = <Main/*firstTarget*/Button />;  // We get undefined for resolved signature indicating an error, then just return the first declaration
+            const {symbolName, symbolKind, containerName}  = getSymbolInfo(typeChecker, symbol, node);
+            return [createDefinitionInfo(symbol.valueDeclaration, symbolKind, symbolName, containerName)];
+        }
+
+        // If the current location we want to find its definition is in an object literal, try to get the contextual type for the
+        // object literal, lookup the property symbol in the contextual type, and use this for goto-definition.
+        // For example
+        //      interface Props{
+        //          /first*/prop1: number
+        //          prop2: boolean
+        //      }
+        //      function Foo(arg: Props) {}
+        //      Foo( { pr/*1*/op1: 10, prop2: true })
+        const container = getContainingObjectLiteralElement(node);
+        if (container) {
+            const contextualType = typeChecker.getContextualType(node.parent.parent as Expression);
+            if (contextualType) {
+                let result: DefinitionInfo[] = [];
+                forEach(getPropertySymbolsFromContextualType(typeChecker, container), contextualSymbol => {
+                    result = result.concat(getDefinitionFromSymbol(typeChecker, contextualSymbol, node));

```

Comment:
I am not sure I understand "just the one that matches  but if i understand correctly, the function `getPropertySymbolsFromContextualType`  done the filtering already

---
In reply to: [87305445](https://github.com/Microsoft/TypeScript/pull/12107#discussion_r87305445) [](ancestors = 87305445)